rxbindings/src/main/java/com/amplifyframework/rx/RxAdapters.java

```diff
@@ -100,4 +103,85 @@ private RxAdapters() {}
+     * Interface that defines a contract for callback-style operations to be
+     * mapped to reactive-style operation types that offer a more idiomatic way
+     * of using reactive programming with the Amplify framework.
+     * @param <T> The type that represents the result of a given operation.
+     * @param <P> The type that represents the progress of a given operation.
+     * @param <E> The type of exception of a given operation.
+     */
+    interface RxProgressAwareCallbackMapper<T, P, E> {
+        Cancelable emitTo(Consumer<P> onProgress, Consumer<T> onItem, Consumer<E> onError);
+    }
+
+    /**
+     * Interface that should be implemented by reactive-style operations
+     * wishing to return a {@link Single} as its result.
+     * @param <T> The type that represents the result of a given operation.
+     */
+    interface RxSingleOperation<T> extends Cancelable {
+        /**
+         * Maps the result of a callback-style operation to a {@link Single}.
+         * @return A {@link Single} that emits a result or an error.
+         */
+        Single<T> observeResult();
+    }
+
+    /**
+     * Interface that should be implemented by reactive-style operations
+     * that emit progress information during processing.
+     * @param <P> The type that represents the progress of a given operation.
+     */
+    interface RxProgressAwareOperation<P> {
+        /**
+         * Returns an observable that the caller can subscribe to
+         * receive progress information about the operation.
+         * @return An observable that emits progress information.
+         */
+        Observable<P> observeProgress();
+    }
+
+    /**
+     * A generic implementation of an operation that emits
+     * progress information and returns a single.
+     * @param <T> The type that represents the result of a given operation.
+     * @param <P> The type that represents the progress of a given operation.
+     */
+    static class RxProgressAwareSingle<T, P>
+        implements RxAdapters.RxSingleOperation<T>, RxProgressAwareOperation<P> {
+
+        private PublishSubject<P> progressSubject;
+        private ReplaySubject<T> resultSubject;
+        private Cancelable amplifyOperation;
+
+        RxProgressAwareSingle(RxProgressAwareCallbackMapper<T, P, ? extends AmplifyException> callbacks) {
+            progressSubject = PublishSubject.create();
+            resultSubject = ReplaySubject.create();
+            amplifyOperation = callbacks.emitTo(progressSubject::onNext,
+                                                resultSubject::onNext,
+                                                resultSubject::onError);
+        }
+
+        @Override
+        public void cancel() {
+            amplifyOperation.cancel();
+            resultSubject.onComplete();
+            progressSubject.onComplete();
+        }
+
+        @Override
+        public Single<T> observeResult() {
+            Single<T> objectSingle = Single.create(emitter -> {
```

Comment:
Just `return Single` without caching to variable?

Reply:
Doh 🤦